documentation/optee_design.md

```diff
@@ -487,6 +487,107 @@ implementation, as explained in [crypto.md].
+`core/arch/arm/sta/stats.c` as a template and just add your static TA based on
+that to the `sub.mk` in the same directory.
+
+###Trusted Applications
+
+Trusted Applications (TAs) are applications dynamically loaded by OP-TEE
+core in the Secure World when something in the REE wants to talk to that
+particular application UUID.  It's similar to the way the Linux
+kernel can dynamically load kernel modules.
+
+Because the TAs are signed by the same key that built the OP-TEE core, they are
+able to be stored in the untrusted REE filesystem, and tee-supplicant will
+take care of passing them to be checked and loaded by the Secure World OP-TEE
+core.  Again this is simular to Linux kernel module signature checking.
+
+Trusted Application benefit from the `GP Core Internal API` as specified by
+the 'GP Specs'. 
+
+Trusted Application consist of a cleartext signed ELF file, named from the UUID
+of the TA and the suffix ".ta".
+
+They are built separately from the optee_os boot-time blob, although when they
+are built they use the same build system, and are signed with the key from the
+build of the original optee_os blob.
+
+## Special treatment of Trusted Applications
+
+### Syscalls
+
+Since (dynamically loaded) TAs are not directly bound to function exports in
+the optee_os blob, since that would require upgrades of all TAs synchronusly to
+upgrades of the optee_os blob, the use of optee_os exports by the TA must be
+resolved at runtime.
+
+optee_os does this by using syscalls, the same kind of way as the Linux kernel
+provides a stable API for its userland programs.  TAs are written to use
+syscall wrappers to access functions exported from optee_os, so this all
+happens automatically when a TA wants to use an api exported from core optee_os.
+
+Static / Psuedo TAs and anything else in the core optee_os blob do not require
+going through a syscall interface, since they can just link directly as they are
+part of the core blob.
+
+Syscalls are provided already for all public exports from the core that a
+Dynamic TA is expected to use, so you only need to take care about this if
+you will add new exported from the core that TAs will want to use.
+
+### Malloc mapping
+
+When the core optee_os performs a `malloc()`, the VA mapping is placed in the
+MMC table only for core optee_os threads.  This means that TAs cannot observe
+data that is being allocated and used by the core optee_os.
+
+When a TA calls TEE_Malloc(), the mapping for the allocation is visible to
+both the core optee_os code and the specific individual TA only.  This enforces
+"Chinese Walls" between the TA views of Secure World.
+
+(This has an implication is you write a syscall handler that does a malloc()
+in optee_os core context on behalf og a TA... there is no mapping in the
+calling TA MMU view for it to actually use the allocation when the syscall
+returns).
+
+### Malloc pool
+
+The core_os malloc heap is defined by `CFG_CORE_HEAP_SIZE` in `mk/config.mk`.
```

Comment:
I agree... however I settled for changing everything referring to the "core OS" in the new section to "OP-TEE OS" and  "code" or "blob" after where it makes sense.